### PR TITLE
fix(message overrides): extract message content with Platform-Specific Overrides

### PR DESCRIPTION
**Pull request by buddyeorl**
#### Description of changes

- Exposed getClientInfo function from @aws-amplify/core
- Refactored the extractContent function to accept a platform argument and use it to extract the correct platform override button actions and links from the message.

#### Issue #, if available
[Issue 12915](https://github.com/aws-amplify/amplify-js/issues/12915)

#### Description of how you validated changes
- Set up pinpoint in app messages with android, and ios and web overrides
- Used Verdaccio to install updated packages on test apps, react native android, ios and react app.
- Confirmed that overrides where extracted correctly and the button actions triggered the correct event.
- Repeated the steps setting up pinpoint in app messages, with no overrides and with only single platform overrides.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/notifications/__tests__/inAppMessaging/utils/processInAppMessages.test.ts [hunk 1]
```diff
@@ -19,7 +19,17 @@ import {
 import { initializeInAppMessaging } from '../../../src/inAppMessaging/providers/pinpoint/apis';
 
 jest.mock('@aws-amplify/core');
-jest.mock('@aws-amplify/core/internals/utils');
+jest.mock('@aws-amplify/core/internals/utils', () => {
```
Comment (Samaritan1011001) [inline packages/notifications/__tests__/inAppMessaging/utils/processInAppMessages.test.ts:22]: Q: Unclear on why this change is required and why it is returning `android`.

**Review by Samaritan1011001 — COMMENTED** [review]
@buddyeorl, appreciate this very much. I added a few questions and suggestions. Looking forward to work with you to get this in.

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts [hunk 2]
```diff
@@ -252,8 +256,22 @@ export const extractContent = ({
 				PrimaryBtn,
 				SecondaryBtn,
 			} = content;
-			const defaultPrimaryButton = PrimaryBtn?.DefaultConfig;
-			const defaultSecondaryButton = SecondaryBtn?.DefaultConfig;
+			// Determine the button configuration based on the platform config or use default
+			const getButtonConfig = (button: any, defaultConfig: any) => {
```
[the remark this answers is not in the document]
Comment (Samaritan1011001) [reply packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts:260]: Function can also be further simplified to 
```ts
const getButtonConfig = (button: any) => ({
	...button?.defaultConfig, ...button?.[configPlatform]
})
```

File: packages/notifications/__tests__/testUtils/data.ts [hunk 3]
```diff
@@ -295,3 +740,23 @@ export const completionHandlerId = 'completion-handler-id';
 export const userAgentValue = 'user-agent-value';
 
 export const channelType = 'APNS_SANDBOX';
+
+export const browserConfigTestCases = [
+	{ os: 'android', expectedPlatform: 'Web' },
+	{ os: 'ios', expectedPlatform: 'Web' },
```
[the remark this answers is not in the document]
Comment (buddyeorl) [reply packages/notifications/__tests__/testUtils/data.ts:649]: when executed within a mobile browser, such as Safari on an iOS device, the `getClientInfo` method could indeed return ios for the platform. However, for our purposes, even in such cases, the expected platform in the configuration should  be marked as web to reflect the web environment, same for mobile browsers on android devices.

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (buddyeorl) [reply packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts]: Simplified getButtonConfig function and added logic to handle undefined button config case [here](https://github.com/aws-amplify/amplify-js/pull/12917/files#diff-5b30aa431c0b57dcc5ca8f31aed06516326bb3b87963cb22543589484fe1f6e1)

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts [hunk 4]
```diff
@@ -252,8 +256,22 @@ export const extractContent = ({
...
+				return configPlatform && button && button[configPlatform]
```
[the remark this answers is not in the document]
Comment (buddyeorl) [reply packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts]: I agree, moved logic [here](https://github.com/aws-amplify/amplify-js/pull/12917/files#diff-5b30aa431c0b57dcc5ca8f31aed06516326bb3b87963cb22543589484fe1f6e1)

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts [hunk 5]
```diff
@@ -252,8 +256,22 @@ export const extractContent = ({
...
+					? { ...defaultConfig, ...button[configPlatform] }
+					: defaultConfig;
+			};
+
+			const defaultPrimaryButton = getButtonConfig(
+				PrimaryBtn,
+				PrimaryBtn?.DefaultConfig
```
[the remark this answers is not in the document]
Comment (buddyeorl) [reply packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts]: Fixed [here](https://github.com/aws-amplify/amplify-js/pull/12917/files#diff-5b30aa431c0b57dcc5ca8f31aed06516326bb3b87963cb22543589484fe1f6e1)

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/messageProcessingHelpers.ts [hunk 6]
```diff
@@ -86,10 +87,15 @@ export async function incrementMessageCounts(messageId: string): Promise<void> {
 function normalizeMessages(messages: PinpointInAppMessage[]): InAppMessage[] {
 	return messages.map(message => {
 		const { CampaignId, InAppMessage } = message;
+		const clientInfo = getClientInfo();
```
[the remark this answers is not in the document]
Comment (buddyeorl) [reply packages/notifications/src/inAppMessaging/providers/pinpoint/utils/messageProcessingHelpers.ts]: Fixed [here](https://github.com/aws-amplify/amplify-js/pull/12917/files#diff-5b30aa431c0b57dcc5ca8f31aed06516326bb3b87963cb22543589484fe1f6e1)

File: packages/notifications/src/inAppMessaging/types/message.ts [hunk 7]
```diff
@@ -13,6 +13,8 @@ export type InAppMessageAction = 'CLOSE' | 'DEEP_LINK' | 'LINK';
 
 export type InAppMessageTextAlign = 'center' | 'left' | 'right';
 
+export type ConfigPlatformType = 'Android' | 'IOS' | 'Web' | 'DefaultConfig';
```
[the remark this answers is not in the document]
Comment (buddyeorl) [reply packages/notifications/src/inAppMessaging/types/message.ts]: fixed here [packages/notifications/src/inAppMessaging/types/message.ts](https://github.com/aws-amplify/amplify-js/pull/12917/files#diff-90a1a0ef46f64eb8e4659f58b14aaffde93fdf95824f5256ddcd18730166a98c)

File: packages/notifications/__tests__/inAppMessaging/providers/pinpoint/utils/helpers.test.ts [hunk 8]
```diff
@@ -271,4 +275,49 @@ describe('InAppMessaging Provider Utils', () => {
...
+				});
+			});
+		});
+
+		describe('when running natively', () => {
+			let originalWindow = window;
+
+			beforeEach(() => {
+				//remove window object to mock behavior outside a browser
+				Object.defineProperty(global, 'window', {});
```
[the remark this answers is not in the document]
Comment (buddyeorl) [reply packages/notifications/__tests__/inAppMessaging/providers/pinpoint/utils/helpers.test.ts]: To avoid affecting other tests, I've separated browser and native tests into different files. For native tests, I used a Jest DocBlock:

``` js
/**
 * @jest-environment node
 */
```
 to switch to Node, bypassing the need to manipulate the window object, thus ensuring the tests remain isolated and specific to their environments.

File: packages/notifications/__tests__/inAppMessaging/providers/pinpoint/utils/helpers.test.ts [hunk 9]
```diff
@@ -271,4 +275,49 @@ describe('InAppMessaging Provider Utils', () => {
...
+			});
+
+			afterEach(() => {
+				//remove window object to mock behavior outside a browser
```
[the remark this answers is not in the document]
Comment (buddyeorl) [reply packages/notifications/__tests__/inAppMessaging/providers/pinpoint/utils/helpers.test.ts]: Fixed!

File: packages/notifications/__tests__/inAppMessaging/utils/processInAppMessages.test.ts (hunk 1 above)
Comment (buddyeorl) [reply packages/notifications/__tests__/inAppMessaging/utils/processInAppMessages.test.ts]:
``` js 
jest.mock('@aws-amplify/core/internals/utils', () => {
	const originalModule = jest.requireActual(
		'@aws-amplify/core/internals/utils'
	);
	return {
		...originalModule,
		getClientInfo: jest.fn(), // Setup as a Jest mock function without implementation
	};
});

```

The code above is the latest version, mocks the return value of the getClientInfo function. Since we're not on a real browser or device, getClientInfo may always return undefined. This is fine for testing the `processInAppMessages` function, where undefined (or any other value) triggers a default button config. But for tests requiring specific platform results, we need to mock getClientInfo accordingly.

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts [hunk 10]
```diff
@@ -331,3 +349,19 @@ export const extractMetadata = ({
 	priority: Priority,
 	treatmentId: TreatmentId,
 });
+
+export const mapOSPlatform = (os: string): ConfigPlatformType => {
```
[the remark this answers is not in the document]
Comment (buddyeorl) [reply packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts]: fixed here : [packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts](https://github.com/aws-amplify/amplify-js/pull/12917/files#diff-5b30aa431c0b57dcc5ca8f31aed06516326bb3b87963cb22543589484fe1f6e1)

File: packages/notifications/__tests__/testUtils/data.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Samaritan1011001) [reply packages/notifications/__tests__/testUtils/data.ts:649]: I see, makes sense, thanks for the explanation.

File: packages/notifications/__tests__/testUtils/data.ts [hunk 11]
```diff
@@ -183,6 +184,353 @@ export const pinpointInAppMessage: PinpointInAppMessage = {
 	TreatmentId: 'T1',
 };
 
+export const browserPinpointInAppMessagesWithOverrides: {
```
Comment (Samaritan1011001) [inline packages/notifications/__tests__/testUtils/data.ts:187]: Nit: Most of the redundant data here can be extracted to a common object and reused.

File: packages/core/src/index.ts [hunk 12]
```diff
@@ -60,7 +60,7 @@ export {
 	sessionStorage,
 	sharedInMemoryStorage,
 } from './storage';
-export { KeyValueStorageInterface } from './types';
+export { KeyValueStorageInterface, InAppMessageButton } from './types';
```
Comment (Samaritan1011001) [inline packages/core/src/index.ts:63]: Sorry, this type needs to be exported internally from [awsClients/pinpoint.index.ts](https://github.com/buddyeorl/amplify-js/blob/290a55fdba7da3072098d8abdc7742128b3205c0/packages/core/src/awsClients/pinpoint/index.ts#L15) itself. 

You can then import it from [@aws-amplify/core/internals/aws-clients/pinpoint](https://github.com/aws-amplify/amplify-js/pull/12917/files#diff-5b30aa431c0b57dcc5ca8f31aed06516326bb3b87963cb22543589484fe1f6e1R9) in the helpers.ts impl.

File: packages/core/src/types/core.ts [hunk 13]
```diff
@@ -49,3 +49,5 @@ export type DelayFunction = (
 	args?: any[],
 	error?: unknown
 ) => number | false;
+
+export { InAppMessageButton } from "../awsClients/pinpoint/types";
```
Comment (Samaritan1011001) [inline packages/core/src/types/core.ts:53]: With the above comment, this won't be necessary here.

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts [hunk 14]
```diff
@@ -1,8 +1,11 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
 
-import { ConsoleLogger } from '@aws-amplify/core';
-import { InAppMessagingAction } from '@aws-amplify/core/internals/utils';
+import { ConsoleLogger, InAppMessageButton } from '@aws-amplify/core';
```
Comment (Samaritan1011001) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts:4]: With the export update in core, you won't need this change. You'll then be importing this on Line 9.

**Review by Samaritan1011001 — COMMENTED** [review]
Thank you for the changes. One more important `core` export update needed but otherwise it looks good to me!

File: packages/notifications/__tests__/testUtils/data.ts (hunk 11 above)
Comment (buddyeorl) [reply packages/notifications/__tests__/testUtils/data.ts]: Makes sense. I've simplified the override button objects and added a couple of utility functions for merging:[mergeInAppMessageWithOverrides.ts](https://github.com/aws-amplify/amplify-js/pull/12917/commits/8c56fe453a501179f477c2fe5fbb1a5ff28d0695#diff-c69a44bf12a466c3e160da3b9f57079e9480cdfb4a97cb7014ec05cac90018af) and here [/testUtils/data.ts](https://github.com/aws-amplify/amplify-js/pull/12917/commits/8c56fe453a501179f477c2fe5fbb1a5ff28d0695#diff-053cb495ee52e974849da7f2c965264adfbd8660c45251d3e349106f7ec0bc91)

 I've also updated the corresponding tests accordingly.

File: packages/core/src/index.ts (hunk 12 above)
Comment (buddyeorl) [reply packages/core/src/index.ts]: No worries, I've updated the export in :[packages/core/src/awsClients/pinpoint/index.ts](https://github.com/aws-amplify/amplify-js/pull/12917/files#diff-2b118443727f81202a936813b7aac79a475f9c835e404f1921d09c367aca2b51),

Additionally, I'm exporting `OverrideButtonConfiguration` for type checking the updated test data and the new merging utility function.

File: packages/core/src/types/core.ts (hunk 13 above)
Comment (buddyeorl) [reply packages/core/src/types/core.ts]: fixed.

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts (hunk 14 above)
Comment (buddyeorl) [reply packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts]: fixed.

**Review by Samaritan1011001 — APPROVED** [review]
Changes look good to me. 
@buddyeorl Thank you for the contribution.

File: packages/notifications/__tests__/testUtils/mergeInAppMessageWithOverrides.ts [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+export const mergeInAppMessageWithOverrides:(
+	pinpointInAppMessage:InAppMessageCampaign,
+	platform:'android'|'ios'|'web',
+	buttonOverrides?:{
+		primaryButton:OverrideButtonConfiguration,
+		secondaryButton:OverrideButtonConfiguration
+	}
+ ) =>InAppMessageCampaign = (pinpointInAppMessage, platform, buttonOverrides)=>{
+	const message= cloneDeep(pinpointInAppMessage);
+	const configPlatform = mapOSPlatform(platform);
```
Comment (cshfang) [inline packages/notifications/__tests__/testUtils/mergeInAppMessageWithOverrides.ts:18]: Would highly prefer not relying on other utils inside a test util as it could be hard to debug if something upstream changes and causes test failures downstream. In this case, since the goal is to augment some `InAppMessageCampaign`'s button configuration with some `ButtonConfigPlatform` keyed values, I think we can just have the function accept that directly. We can also tidy up the signature a bit by typing the parameters inline: E.g.
```suggestion
export const mergeInAppMessageWithOverrides = (
	pinpointInAppMessage: InAppMessageCampaign,
	configPlatform: ButtonConfigPlatform,
	buttonOverrides?: {
		primaryButton: OverrideButtonConfiguration;
		secondaryButton: OverrideButtonConfiguration;
	}
): InAppMessageCampaign => {
	const message = cloneDeep(pinpointInAppMessage);
```

File: packages/notifications/__tests__/testUtils/mergeInAppMessageWithOverrides.ts [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+	return message
+}
+
+export const mergeExpectedContentWithExpectedOverride:(
+	inAppMessage:InAppMessageContent,
+	expectedButtonConfig: {
+		primaryButton:OverrideButtonConfiguration,
+		secondaryButton:OverrideButtonConfiguration
+	}
+)=>InAppMessageContent = (inAppMessage, expectedButtonConfig)=>{
```
Comment (cshfang) [inline packages/notifications/__tests__/testUtils/mergeInAppMessageWithOverrides.ts:41]: Nit: Ditto here, could tidy up the typing a bit
```suggestion
export const mergeExpectedContentWithExpectedOverride = (
	inAppMessage: InAppMessageContent,
	expectedButtonConfig: {
		primaryButton: OverrideButtonConfiguration;
		secondaryButton: OverrideButtonConfiguration;
	}
): InAppMessageContent => {
```

File: packages/notifications/__tests__/inAppMessaging/providers/pinpoint/utils/helpers.test.ts [hunk 17]
```diff
@@ -271,4 +285,42 @@ describe('InAppMessaging Provider Utils', () => {
...
+					const expectedContent = mergeExpectedContentWithExpectedOverride(extractedContent[0], buttonOverrides);
+
+					test(`correctly extracts content for ${configPlatform}`, () => {
+						const utils = require('@aws-amplify/core/internals/utils');
+						// Dynamically override the mock for getClientInfo
+						utils.getClientInfo.mockImplementation(() => ({
+							platform: configPlatform,
+						}));
+
+						const content = extractContent(message);
```
Comment (cshfang) [inline packages/notifications/__tests__/inAppMessaging/providers/pinpoint/utils/helpers.test.ts:314]: Nit: Ditto

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts [hunk 18]
```diff
@@ -252,8 +258,15 @@ export const extractContent = ({
 				PrimaryBtn,
 				SecondaryBtn,
 			} = content;
-			const defaultPrimaryButton = PrimaryBtn?.DefaultConfig;
-			const defaultSecondaryButton = SecondaryBtn?.DefaultConfig;
+
+			const getButtonConfig = (button?: InAppMessageButton) => ({
+				...button?.DefaultConfig,
+				...button?.[configPlatform],
+			});
```
Comment (cshfang) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts:265]: Nit: I think there's some redundancy here with the checking given the usage below on lines 267/268. I think this function can also be extracted to the module scope so it's not recreated each time `extractContent` is called. E.g.

```ts
const getButtonConfig = (
	configPlatform: ButtonConfigPlatform,
	button?: InAppMessageButton
): DefaultButtonConfiguration | undefined => {
	if (!button) {
		return;
	}
	return {
		...button?.DefaultConfig,
		...button?.[configPlatform],
	};
};
```

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts [hunk 19]
```diff
@@ -252,8 +258,15 @@ export const extractContent = ({
...
+
+			const defaultPrimaryButton = PrimaryBtn ? getButtonConfig(PrimaryBtn) : undefined;
+			const defaultSecondaryButton = SecondaryBtn ? getButtonConfig(SecondaryBtn) : undefined;
```
Comment (cshfang) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts:268]: Nit: Related to the above comment, the responsibility for checking existence becomes solely the function's
```suggestion
			const defaultPrimaryButton = getButtonConfig(PrimaryBtn);
			const defaultSecondaryButton = getButtonConfig(SecondaryBtn);
```

**Review by cshfang — COMMENTED** [review]
Overall looks pretty close to me. I just have a few suggestions to improve maintainability. I also think there are some formatting issues - I know our repo is currently not set up to auto-format but it would be helpful if you could run these changes through [Prettier](https://prettier.io/) (the repo should have prettier settings configured)

File: packages/notifications/__tests__/inAppMessaging/providers/pinpoint/utils/helpers.test.ts [hunk 20]
```diff
@@ -271,4 +285,42 @@ describe('InAppMessaging Provider Utils', () => {
 
 		expect(extractMetadata(message)).toStrictEqual(extractedMetadata);
 	});
+
+	describe('mapOSPlatform method', () => {
+		describe('when running in a browser', () => {
```
[the remark this answers is not in the document]
Comment (buddyeorl) [reply packages/notifications/__tests__/inAppMessaging/providers/pinpoint/utils/helpers.test.ts]: I've flattened the tests for this one. However, I'm not updating the suite description since most tests are applicable to both browser and native environments, despite all tests currently using jsdom. Instead, I prefer to add a flag to specific tests to differentiate between browser and native contexts, keeping the test suite description unchanged to avoid confusion.

[flatten test structure](https://github.com/aws-amplify/amplify-js/pull/12917/commits/d7721c397e1b4e676da8375682c36adfa38f5a68)

File: packages/notifications/__tests__/inAppMessaging/providers/pinpoint/utils/helpers.native.test.ts [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+	);
+	return {
+		...originalModule,
+		getClientInfo: jest.fn(), // Setup as a Jest mock function without implementation
+	};
+});
+
+describe('InAppMessaging Provider Utils', () => {
+	describe('mapOSPlatform method', () => {
+		describe('when running natively', () => {
```
[the remark this answers is not in the document]
Comment (buddyeorl) [reply packages/notifications/__tests__/inAppMessaging/providers/pinpoint/utils/helpers.native.test.ts]: makes sense, fixed here: [flatten test structure](https://github.com/aws-amplify/amplify-js/pull/12917/commits/d7721c397e1b4e676da8375682c36adfa38f5a68)

File: packages/notifications/__tests__/inAppMessaging/providers/pinpoint/utils/helpers.native.test.ts [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+				test(`correctly maps OS "${os}" to ConfigPlatformType "${expectedPlatform}"`, () => {
+					const result = mapOSPlatform(os);
+					expect(result).toBe(expectedPlatform);
+				});
+			});
+		});
+	});
+
+	describe('extractContent with overrides', () => {
+		describe('when running natively', () => {
```
[the remark this answers is not in the document]
Comment (buddyeorl) [reply packages/notifications/__tests__/inAppMessaging/providers/pinpoint/utils/helpers.native.test.ts]: [flatten test structure](https://github.com/aws-amplify/amplify-js/pull/12917/commits/d7721c397e1b4e676da8375682c36adfa38f5a68)

File: packages/notifications/__tests__/inAppMessaging/providers/pinpoint/utils/helpers.native.test.ts [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+							platform: configPlatform,
+						}));
+
+						const content = extractContent(message);
+						expect(content[0].primaryButton).toStrictEqual(
+							expectedContent.primaryButton
+						);
+						expect(content[0].secondaryButton).toStrictEqual(
+							expectedContent.secondaryButton
+						);
```
[the remark this answers is not in the document]
Comment (buddyeorl) [reply packages/notifications/__tests__/inAppMessaging/providers/pinpoint/utils/helpers.native.test.ts]: updated here:
[minor update](https://github.com/aws-amplify/amplify-js/pull/12917/commits/b53cefc411391c958ac58838b2e1cc58218310fc)

File: packages/notifications/__tests__/inAppMessaging/providers/pinpoint/utils/helpers.test.ts (hunk 17 above)
Comment (buddyeorl) [reply packages/notifications/__tests__/inAppMessaging/providers/pinpoint/utils/helpers.test.ts]: updated here:
[minor update](https://github.com/aws-amplify/amplify-js/pull/12917/commits/59828bd1b19afbe3a0e0bcb961c8f8a22a43bab7)

File: packages/notifications/__tests__/testUtils/mergeInAppMessageWithOverrides.ts (hunk 15 above)
Comment (buddyeorl) [reply packages/notifications/__tests__/testUtils/mergeInAppMessageWithOverrides.ts]: I agree, but I see minimal risk in debugging, since I've added a test for the mappedOS utils. If any upstream changes affect this function, the test should fail, highlighting the issue. Regardless, to avoid using additional utils in the test utils, I updated the mock data to include a mappedPlatform property alongside configPlatform and passed it down. I'm open to suggestions.

[updated merge override fns signature and associated tests](https://github.com/aws-amplify/amplify-js/pull/12917/commits/3a58f157071dcb7ee62a9e8bf062c96bb24d7cd0)

File: packages/notifications/__tests__/testUtils/mergeInAppMessageWithOverrides.ts (hunk 16 above)
Comment (buddyeorl) [reply packages/notifications/__tests__/testUtils/mergeInAppMessageWithOverrides.ts]: updated here:
[updated merge override fns signature and associated tests](https://github.com/aws-amplify/amplify-js/pull/12917/commits/3a58f157071dcb7ee62a9e8bf062c96bb24d7cd0)

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts (hunk 18 above)
Comment (buddyeorl) [reply packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts]: removed redundancy, extracted function and added additional checks:
[extracted getButtonConfig fn and refactored to add data existence che…](https://github.com/aws-amplify/amplify-js/pull/12917/commits/7ff8b96827cccda8f1a14598f8f426652cfbb879)

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts (hunk 19 above)
Comment (buddyeorl) [reply packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts]: updated here:
[extracted getButtonConfig fn and refactored to add data existence che…](https://github.com/aws-amplify/amplify-js/pull/12917/commits/7ff8b96827cccda8f1a14598f8f426652cfbb879)

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts [hunk 24]
```diff
@@ -368,3 +362,19 @@ export const mapOSPlatform = (os?: string): ButtonConfigPlatform => {
 			return 'DefaultConfig';
 	}
 };
+
+const getButtonConfig = (
+	configPlatform: ButtonConfigPlatform,
+	button?: InAppMessageButton,
+): InAppMessageButton['DefaultConfig'] | undefined => {
+	if (!button || !button?.DefaultConfig) {
```
Comment (cshfang) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts:370]: Nit: I think this is effectively the same as
```ts
if (!button?.DefaultConfig) {
```

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts [hunk 25]
```diff
@@ -368,3 +362,19 @@ export const mapOSPlatform = (os?: string): ButtonConfigPlatform => {
...
+		return;
+	}
+	if (!configPlatform || !button?.[configPlatform]) {
+		return button?.DefaultConfig;
+	}
+	return {
+		...button?.DefaultConfig,
+		...button?.[configPlatform],
+	};
```
Comment (cshfang) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts:379]: Given the guard clause above and how the spread operator works, I think we can actually condense this:
```ts
return {
  button.DefaultConfig,
  button.configPlatform,
}
```
Since we know `button` is guaranteed to be truthy by this point and even if `configPlatform` is nullish, it would spread just fine and we would end up returning the DefaultConfig.

**Review by cshfang — COMMENTED** [review]
One last small change and I think we're probably good to go. Thanks for bearing with me on this

**Comment by cshfang** [comment]
> > Overall looks pretty close to me. I just have a few suggestions to improve maintainability. I also think there are some formatting issues - I know our repo is currently not set up to auto-format but it would be helpful if you could run these changes through [Prettier](https://prettier.io/) (the repo should have prettier settings configured)
> 
> @cshfang, Oops, I'm on a new machine and realized I hadn't set up any plugins in my IDE. I've now added Prettier and reformatted all the modified/added code. It should be good now: [prettier](https://github.com/aws-amplify/amplify-js/pull/12917/commits/a7d9f4309984b36bc4d1c7612ba849494e84617c)

No worries - really appreciate it

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts (hunk 24 above)
Comment (buddyeorl) [reply packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts]: Yes, make sense to me:

[removed unnecessary optional chaining operators](https://github.com/aws-amplify/amplify-js/pull/12917/commits/f7eb8005fb0b44f966477bc1cc26d1bf77153137)

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts (hunk 25 above)
Comment (buddyeorl) [reply packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts]: updated here:
[removed unnecessary optional chaining operators](https://github.com/aws-amplify/amplify-js/pull/12917/commits/f7eb8005fb0b44f966477bc1cc26d1bf77153137)

**Review by cshfang — DISMISSED** [review]

**Review by Samaritan1011001 — APPROVED** [review]

**Comment by Samaritan1011001** [comment]
@buddyeorl Apologies for the delay here. I was finally able to verify the changes on both Android and iOS devices work properly. However, there are a few lint fixes you'll have to make before we merge. The failing test should give you the info needed. Looking to merge once these are updated.

**Comment by buddyeorl** [comment]
> @buddyeorl Apologies for the delay here. I was finally able to verify the changes on both Android and iOS devices work properly. However, there are a few lint fixes you'll have to make before we merge. The failing test should give you the info needed. Looking to merge once these are updated.

@Samaritan1011001 Thanks for the update! I've addressed the lint fixes and re-run the tests on my end. Everything should be passing smoothly now. Please let me know if there's anything else needed from my side. Appreciate your assistance!

**Review by Samaritan1011001 — APPROVED** [review]

**Review by cshfang — APPROVED** [review]
